psi4/src/psi4/libmints/mcmurchiedavidson.cc

```diff
@@ -160,4 +160,78 @@ void fill_M_matrix(int maxam, int maxpow, const Point& PC, double a, double b, s
     }
 }
 
+void fill_R_matrix(int maxam, double p, const Point& P, const Point& C, std::vector<double>& R,
+                   std::shared_ptr<const libint2::FmEval_Chebyshev7<double>> fm_eval) {
+    // Generates the auxiliary integrals for Coulomb-type integrals using eq 9.9.13
+    // from Molecular Electronic-Structure Theory (10.1002/9781119019572)
+    auto PC = point_diff(P, C);
+    auto RPC = point_norm(PC);
+    double T = p * RPC * RPC;
+
+    std::vector<double> fmvals(maxam + 1);
+
+    // evaluate Boys function
+    fm_eval->eval(fmvals.data(), T, maxam);
+
+    int dim1 = maxam + 1;
+    int dim2 = dim1 * dim1 * dim1;
+    // R matrix buffer size needs to be at least dim1 * dim2,
+    // only zero out the required part of the buffer for performance
+    std::fill(R.begin(), R.begin() + dim1 * dim2, 0.0);
+
+    // NOTE: avoiding std::pow(-2.0 * p, n)
+    double fac = 1.0;
+    double mult = -2.0 * p;
+    for (int n = 0; n < dim1; ++n) {
+        // eq 9.9.14
+        R[n * dim2] = fac * fmvals[n];
+        fac *= mult;
+    }
+    // t + u + v <= N
+    // t = 0, u = 0
+    for (int v = 1; v < dim1; ++v) {
+        for (int n = 0; n < maxam; ++n) {
+            double val = 0.0;
+            int noffset = (n + 1) * dim2;
+            // eq 9.9.20
+            if (v > 1) {
+                val += (v - 1) * R[noffset + v - 2];  // R_{0,0,v-2}^{n+1}
+            }
+            val += PC[2] * R[noffset + v - 1];  // R_{0,0,v-1}^{n+1}
+            R[n * dim2 + v] = val;              // R_{0,0,v}^{n}
+        }
+    }
+    // t = 0
+    for (int v = 0; v < dim1; ++v) {
+        for (int u = 1; u < dim1 - v; ++u) {
+            for (int n = 0; n < maxam; ++n) {
+                double val = 0.0;
+                int noffset = (n + 1) * dim2;
+                // eq 9.9.19
+                if (u > 1) {
+                    val += (u - 1) * R[noffset + (u - 2) * dim1 + v];  // R_{0,u-2,v}^{n+1}
+                }
+                val += PC[1] * R[noffset + (u - 1) * dim1 + v];  // R_{0,u-1,v}^{n+1}
+                R[n * dim2 + u * dim1 + v] = val;                // R_{0,u,v}^{n}
+            }
+        }
+    }
+    for (int v = 0; v < dim1; ++v) {
+        for (int u = 0; u < dim1 - v; ++u) {
+            for (int t = 1; t < dim1 - v - u; ++t) {
+                for (int n = 0; n < maxam; ++n) {
+                    double val = 0.0;
+                    int noffset = (n + 1) * dim2;
+                    // eq 9.9.18
+                    if (t > 1) {
+                        val += (t - 1) * R[noffset + address_3d(t - 2, u, v, dim1, dim1)];  // R_{t-2,u,v}^{n+1}
+                    }
+                    val += PC[0] * R[noffset + address_3d(t - 1, u, v, dim1, dim1)];  // R_{t-1,u,v}^{n+1}
+                    R[n * dim2 + address_3d(t, u, v, dim1, dim1)] = val;              // R_{t,u,v}^{n}
+                }
+            }
+        }
+    }
+}
```

Comment:
Is this the most expensive part of the multipole potential integral calculation? If so, I wonder if there'd be any performance gains from flattening these loops.

Reply:
What exactly do you mean by "flattening" in this case? Yes, the snippet you mention is one of the hotspots, but gathering all the contributions [here](https://github.com/psi4/psi4/blob/037ee51b2726f38f5fa79d258f1d05a223a78b7e/psi4/src/psi4/libmints/multipolepotential.cc#L148) is even more expensive.

Reply:
I've tested the flattened loop version, which is unfortunately much slower than the 3-loop in this case, but it was worth a try I guess 😄

Reply:
Haha, thanks for giving it a shot! Holger is definitely right about integer division